### PR TITLE
Doc Update on Heter. Systems

### PR DESCRIPTION
**Pull request by abejgonzalez**
Explains how to use MultiRoCC and create multi-core systems with both BOOM and RoCC. Also creates a 'Advanced-Usage' section to hold this.

### CODE REVIEW

**Comment by abejgonzalez** [comment]
Note: This is rebased on #125 so it must be merged after that PR.

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      new WithBootROM ++
+      new boom.system.WithRenumberHarts ++
+      new boom.common.WithRVC ++
+      new boom.common.DefaultBoomConfig ++
+      new boom.system.WithNBoomCores(2) ++
+      new freechips.rocketchip.subsystem.WithoutTLMonitors ++
+      new freechips.rocketchip.subsystem.WithNBigCores(1) ++
+      new freechips.rocketchip.system.BaseConfig)
+
+In this example, the ``WithN...Cores(X)`` mixins set up particular parameters for the duplicated core.
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: This sentence might be a bit confusing. Does this imply default parameters? If so, why do you need the additional (DefaultBOOMConfig mixin)?

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+For example, the Rocket tiles will be "BigCores" while BOOM has some generic parameters set in its mixin.
+In BOOMs case, we override these base core parameters to create multiple "DefaultBoom"s with calling ``DefaultBoomConfig`` after the multi-core instantiation.
+This mixin applies to all the BOOM cores in the system and sets up the parameters for each.
+
+Great! Now you have a heterogeneous setup with BOOMs and Rockets.
+The final thing you need to make this system work is to renumber the ``hartId``'s of the cores so that each core has a unique ``hartId``.
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: Are we assuming we explained ``hartId``s somewhere else in the documentation?

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+.. code-block:: scala
+
+    class SixCoreConfig extends Config(
+      new WithNormalBoomRocketTop ++
+      new WithBootROM ++
+      new WithHeterCoresSetup ++
+      new freechips.rocketchip.system.BaseConfig)
+
+Note, in this setup you most likely don't need the ``WithRenumberHarts`` mixin since you have to assign the ``hartId`` of each tile in the tile parameters yourself.
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: "most likely"?

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      new freechips.rocketchip.system.BaseConfig)
+
+In this example, the ``WithRenumberHarts`` relabels the ``hartId``'s of all the BOOM/Rocket cores.
+Then after that is applied to the parameters, the ``WithMultiRoCCHwacha(0)`` is used to assign to ``hartId`` zero a Hwacha (in this case ``hartId`` zero is Rocket).
+Finally, the ``WithMultiRoCC`` mixin is called.
+This mixin sets the ``BuildRoCC`` key to use the ``MultiRoCCKey`` instead of the default.
+This must be used after all the RoCC parameters are set because it needs to override the ``BuildRoCC`` parameter.
+If this is used earlier in the configuration sequence, then MultiRoCC does not work.
+
+This mixin can be changed to put more accelerators on more cores by changing the arguments to cover more ``hartId``'s (i.e. ``WithMultiRoCCHwacha(0,1,3,6,...)``).
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: Should we mention a the generator approach here if we don't want to hard-code a hartId? (use length or number of tiles to apply the accelerator only to the last tile, or something of that sort)

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst (hunk 1 above)
Comment (abejgonzalez) [reply docs/Advanced-Usage/Heterogeneous-SoCs.rst]: Hmm. So this is just changed in BOOM land. Now, you have to have both mixins (aka. the With... and the XTypeConfig). I can reword this to something like: "In this example, the with "WithN...Cores(Z)" sets up the default parameters needed to have Z amount of cores. For something like BOOM, we add an extra mixin called "XYZ" that is used to change the parameters to a different size core."

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst (hunk 2 above)
Comment (abejgonzalez) [reply docs/Advanced-Usage/Heterogeneous-SoCs.rst]: I'm assuming that `hartId` is known by the person using the core. It is in the spec, however, I can add a sentence if you think it is necessary.

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst (hunk 3 above)
Comment (abejgonzalez) [reply docs/Advanced-Usage/Heterogeneous-SoCs.rst]: So.... if they take the default it will be broken because all cores with get `hartId == 0`. So they will have to set the `hartId`. I will make this clearer.

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst (hunk 4 above)
Comment (abejgonzalez) [reply docs/Advanced-Usage/Heterogeneous-SoCs.rst]: If I understand you correctly, to do your proposal you need a bunch of extra mixins (WithLastAccel, WithBeginningAccel, etc). IMHO, this is a cleaner approach because it makes sure that the user knows what accel. is attached to what system.

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      new freechips.rocketchip.subsystem.WithoutTLMonitors ++
+      new freechips.rocketchip.subsystem.WithNBigCores(1) ++
+      new freechips.rocketchip.system.BaseConfig)
+
+In this example, the ``WithNBoomCores`` and ``WithNBigCores`` mixins set up the default parameters for the multiple BOOM and Rocket cores, respectively.
+However, for BOOM, an extra mixin called ``DefaultBoomConfig`` is added to override the default parameters with a different set of more common default parameters.
+This mixin applies to all BOOM cores in the system and changes the parameters for each.
+
+Great! Now you have a heterogeneous setup with BOOMs and Rockets.
+The final thing you need to make this system work is to renumber the ``hartId``'s of the cores so that each core has a unique ``hartId`` (a ``hartId`` is the hardware thread id of the core).
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: works

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+This is done with ``WithRenumberHarts`` (which can label the Rocket cores first or the BOOM cores first).
+The reason this is needed is because by default the ``WithN...Cores(X)`` mixin assumes that there are only BOOM or only Rocket cores in the system.
+Thus, without the ``WithRenumberHarts`` mixin, each set of cores is labeled starting from zero causing multiple cores to be assigned the same ``hartId``.
+
+Another option to create a multi heterogeneous core system is to override the parameters yourself so you can specify the core parameters per core.
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: Maybe we should emphasize that this is an alternative, and not an addition to the previous method? (I think it's clear, but it probably can't hurt to add additional separation?)

File: docs/Advanced-Usage/Heterogeneous-SoCs.rst [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      new freechips.rocketchip.subsystem.WithoutTLMonitors ++
+      new freechips.rocketchip.subsystem.WithNBigCores(1) ++
+      new freechips.rocketchip.system.BaseConfig)
+
+In this example, Hwachas are added to both BOOM tiles and to the Rocket tile.
+All with the same Hwacha parameters.
+
+Assigning Accelerators to Specific Tiles with MultiRoCC
+-------------------------------------------------------
+
```
Comment (alonamid) [inline docs/Advanced-Usage/Heterogeneous-SoCs.rst]: Maybe mention that in this example Hwacha is an example of a RoCC accelerator?
